src/stdatamodels/stnode.py

```diff
@@ -0,0 +1,565 @@
+    print('schema =', schema)
+    temp_schema.update(schema)
+    validator = asdfschema.get_validator(temp_schema,
+                                          validator_context,
+                                          validator_callbacks,
+                                          validator_resolver)
+
+    #value = yamlutil.custom_tree_to_tagged_tree(value, validator_context)
+    validator.validate(value, _schema=temp_schema)
+    validator_context.close()
+
+def _validate(attr, instance, schema, ctx):
+    tagged_tree = yamlutil.custom_tree_to_tagged_tree(instance, ctx)
+    return _value_change(attr, tagged_tree, schema, False, strict_validation, ctx)
+
+def _get_schema_for_property(schema, attr):
+    subschema = schema.get('properties', {}).get(attr, None)
+    print('XXXX _get_schema_for_property', schema)
+    if subschema is not None:
+        return subschema
+    for combiner in ['allOf', 'anyOf']:
+        for subschema in schema.get(combiner, []):
+            subsubschema = _get_schema_for_property(subschema, attr)
+            if subsubschema != {}:
+                return subsubschema
+    return {}
+
+
+class DNode(UserDict):
+
+    _tag = None
+    _ctx = None
+
+    def __init__(self, node=None, parent=None, name=None):
+
+        if node is None:
+            self.__dict__['_data']= {}
+        elif isinstance(node, dict):
+            self.__dict__['_data'] = node
+        else:
+            raise ValueError("Initializer only accepts dicts")
+        self._x_schema = None
+        self._schema_uri = None
+        self._parent = parent
+        self._name = name
+        # else:
+        #     self.data = node.data
+
+    @property
+    def ctx(self):
+        if self._ctx is None:
+            DNode._ctx = asdf.AsdfFile()
+        return self._ctx
+    
+   
+    def __getattr__(self, key):
+        """
+        Permit accessing dict keys as attributes, assuming they are legal Python
+        variable names.
+        """
+        if key in self._data:
+            print('---------- (getattr) key = ', key)
+            value = self._data[key]
+            if isinstance(value, dict):
+                return DNode(value, parent=self, name=key)
+            elif isinstance(value, list):
+                return LNode(value)
+            else:
+                return value
+        else:
+            raise KeyError(f"No such key ({key}) found in node")
+
+    def __setattr__(self, key, value):
+        """
+        Permit assigning dict keys as attributes.
+        """
+        if key[0] != '_':
+            if key in self._data:
+                print('***** (setattr)', key)
+                if validate:
+                    print('._data', self._data)
+                    self._schema()
+                    schema = self._x_schema.get('properties').get(key, None)
+                    print('schema from __setattr__:', key, schema)
+                    if _validate(key, value, schema, self.ctx):
```

Comment:
Is the intention to eventually allow users to e.g. assign a dict to the jwst.meta.exposure key and automatically convert it to an `Exposure` object?  Or will the dict always fail validation here (due to the schema's tag validator)?

Reply:
Not at the moment. Currently it should require an Exposure object, though eventually I suppose that could be permitted. I would have to think about it.